vllm/model_executor/layers/sampler.py

```diff
@@ -163,19 +163,22 @@ def _apply_min_tokens_penalty(
             start_idx += sampling_metadata.prompt_lens[i] - 1
 
         min_tokens = sampling_params.min_tokens
-        if min_tokens > 0:
+        eos_token_id = sampling_params.eos_token_id
+        stop_token_ids = sampling_params.stop_token_ids
+        if min_tokens > 0 and (eos_token_id is not None or stop_token_ids):
             seqs_to_penalize = []
-            for i, seq_id in enumerate(seq_ids):
+            for j, seq_id in enumerate(seq_ids):
                 seq_data = sampling_metadata.seq_data[seq_id]
                 if len(seq_data.output_token_ids) < min_tokens:
-                    seqs_to_penalize.append(i)
+                    seqs_to_penalize.append(j)
 
             if seqs_to_penalize:
                 # convert to the index into logits
-                seqs_to_penalize = [start_idx + i for i in seqs_to_penalize]
-                # use set() to remove any duplicates
-                token_ids_to_penalize = set(sampling_params.stop_token_ids +
-                                            [sampling_params.eos_token_id])
+                seqs_to_penalize = [start_idx + j for j in seqs_to_penalize]
```

Comment:
wait I don't understand the both `i` -> `j` here

Reply:
@simon-mo there is an outer loop over the sequence groups enumerated with index `i` and within that two separate loops that now each use index `j` (well this one is actually a comprehension)

Reply:
We could change this one to `k` just to be clearer? but they don't overlap so I think it's ok like this too...

Reply:
@simon-mo let me know if you're good with this, I think this is ready to merge if so!